core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -56,7 +60,9 @@ class LogManager(val logDirs: Array[File],
                  scheduler: Scheduler,
                  val brokerState: BrokerState,
                  brokerTopicStats: BrokerTopicStats,
-                 time: Time) extends Logging {
+                 val zkUtils: ZkUtils,
```

Comment:
When `LogCleaner` encounters IOException when cleaning up the log, we want to mark the corresponding log directory as offline and inform controller of the log directory failure via zookeeper. Note that this IOException is not triggered by an external request and thus will not go through `ReplicaManager`. Since we probably don't want to reference ReplicaManager via LogManager, `LogManager.handleLogDirFailure()` seems to be the only reasonable place to have this log of wrting to log directory notification znode.

I couldn't find a better way to address this problem. Do you have a better solution?